Fast-forward merge dashing to main

jacobperron opened the pull request:
@esteve If this looks good to you, I can fast-forward merge `dashing` into `master` from the command line so that the head of each branch is the same.

I'll follow-up with any necessary updates to CI and documentation, as well as PRs for compatibility with Eloquent.

esteve commented on rcljava/include/org_ros2_rcljava_contexts_ContextImpl.h:
```diff
@@ -0,0 +1,59 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
```
I don't know how I feel about this. I'd rather change all the copyright notices (the ones that are right now in master and any future notices) to something generic like "ros2-java developers", instead of adding OSRF. I prefer to keep ros2-java separate from OSRF, as it may lead people to think that it's an OSRF project.

jacobperron replied:
I think if an individual or organization contributes a substantial change to a source file they should be able to add their name to the copyright notice. It can act as an incentive for contributors. For example, this is a common practice we follow for packages in the [ros2](https://github.com/ros2) org. 

Alternatively, if there's preference to add a generic notice to every file, I think it'd be nice to also add a "CONTRIBUTORS" file listing major contributors. Contributors are tracked with version control, but if the system changes or project migrates far in the future at least there's a file record.

I don't know the legal implications of either option. But, my personal preference is to list each copyright owner in every file since it looks less ambiguous.

esteve replied:
> I think if an individual or organization contributes a substantial change to a source file they should be able to add their name to the copyright notice. It can act as an incentive for contributors.

But that's not what copyright noticed are for. They are not for acknowledging contributors, but to state the terms of how the software is distributed.

> For example, this is a common practice we follow for packages in the ros2 org.

That's interesting. What I remember from my days at OSRF is that it respected the copyright notices of the ROS 1 packages, despite having taken over of stewardship of it.

> Alternatively, if there's preference to add a generic notice to every file, I think it'd be nice to also add a "CONTRIBUTORS" file listing major contributors. Of course, individual contributions are tracked with version control, but if the version control system changes or project migrates far in the future at least there's a file record.

I like that more, and there's less arguing about what constitutes a substantial change.

jacobperron replied:
> What I remember from my days at OSRF is that it respected the copyright notices of the ROS 1 packages, despite having taken over of stewardship of it

For ports of ROS 1 packages, we've kept the same license, but often added multiple copyright notices where substantial changes were made (e.g. new features). For example, in RViz: https://github.com/ros2/rviz/blob/5595a69398df4d93aea711d408772ea34c53c230/rviz_common/include/rviz_common/factory/factory.hpp#L2-L4

---

See https://github.com/ros2-java/ros2_java/pull/124 for a proposed change to the copyright notices.

clalancette replied:
> But that's not what copyright noticed are for. They are not for acknowledging contributors, but to state the terms of how the software is distributed.

For what its worth, the style of adding the copyright owner at the top of the license is a very standard practice.  For instance, https://opensource.org/licenses/Apache-2.0 explains on adding the license, a copyright owner should be added.  In most projects I've seen (including ROS 2), the original author of the file is the copyright owner, but I don't have documentation for that, just anecdotal evidence.

The other way I've seen to do it is to make the copyright owner "The FOO Authors", and then list the contributors separately as @jacobperron mentioned (this is how Cartographer [does it](https://github.com/cartographer-project/cartographer/blob/master/cartographer/io/fake_file_writer_test.cc)).

jacobperron replied:
After discussing "offline", we've decided on a policy that new files may contain a different copyright notice while existing files should not have their copyright notice changed or added to.

jacobperron replied:
I found one file that was in contradiction to the policy and fixed it: d270e67

jacobperron commented on README.md:
```diff
@@ -144,7 +144,7 @@ export ANDROID_TOOLCHAIN_NAME=arm-linux-androideabi-clang
 mkdir -p ${ROS2_ANDROID_WORKSPACE}/src
 cd ${ROS2_ANDROID_WORKSPACE}
-wget https://raw.githubusercontent.com/esteve/ros2_java/master/ros2_java_android.repos
+wget https://raw.githubusercontent.com/ros2-java/ros2_java/dashing/ros2_java_android.repos
```
I've been meaning to update the README anyways. https://github.com/ros2-java/ros2_java/pull/119 updates the README for `dashing`. I'll add it to this fast-forward to `master` after it is merged.

jacobperron commented on ros2_java_android.repos:
```diff
@@ -2,80 +2,84 @@ repositories:
     type: git
     url: https://github.com/ros2/ament_cmake_ros.git
-    version: 0.5.0
+    version: dashing
```
That's correct. Though, I planning to make this change *after* the fast-forward. This way the commits on `dashing` and `master` will be even up to the point where we start making changes for Eloquent and beyond.

jacobperron commented on rosidl_generator_java/src/test/java/org/ros2/generator/InterfacesTest.java:
```diff
@@ -34,61 +35,18 @@ public static void setupOnce() {
-    assertEquals('\u007f', rosidl_generator_java.msg.Constants.TOTO);
-    assertEquals(48, rosidl_generator_java.msg.Constants.TATA);
+  // TODO(jacobperron): Replace with JUnit's assertThrows method when we switch to JUnit 5
+  // See: https://junit.org/junit5/docs/5.0.1/api/org/junit/jupiter/api/Assertions.html
```
Issue created https://github.com/ros2-java/ros2_java/issues/123

I haven't checked for compatibility. I can't think of other reasons to switch to JUnit 5 at the moment (besides removing this TODO), so I would say it's low priority.

jacobperron commented on rcljava/src/main/java/org/ros2/rcljava/RCLJava.java:
```diff
@@ -103,52 +127,57 @@ private static void cleanup() {
-        logger.info("Using RMW implementation: {}", RCLJava.getRMWIdentifier());
-        initialized = true;
-      }
+  public static synchronized void rclJavaInit() {
```
I don't think this is true. My understanding is that synchronizing the static method acquires the same class-level lock as the one used in `getGlobalExecutor`. I.e. I believe the following two are equivalent:

```java
public static synchronized void rclJavaInit() {
  ...
}
```

and

```java
public static void rclJavaInit() {
  synchronized (RCLJava.class)  {
    ...
  }
}
```

[Reference](https://docs.oracle.com/javase/tutorial/essential/concurrency/locksync.html)

jacobperron commented:
I've rebased this PR and targeted `main`.
I've also updated the CI configuration to try and make it green (05c98dd).

@ivanpauno Let me know if this batch of changes will make porting things form our fork easier and we can merge it first.

jacobperron commented:
@esteve it would be great if you could take another look at the changes proposed in this PR. It will effectively make `main` compatible with Dashing. Then we can add updates for other distros from there.

ivanpauno commented:
@jacobperron is this a fast-forward from https://github.com/osrf/ros2_java/commits/dashing?

I see some changes in https://github.com/ros2_java/ros2_java/commits/dashing that aren't here.
I also see that there are conflicts between `osrf:dashing` and `osrf:galactic_devel`, I will check what happened there.

---

More info

```
$ git merge-base osrf/dashing origin/dashing
```

cf2ef0707f7420f9b83e6c053775c9d52752a9e2

```
$ git rev-list cf2ef0707f7420f9b83e6c053775c9d52752a9e2^..osrf/dashing
```

cf2ef0707f7420f9b83e6c053775c9d52752a9e2 

\# i.e. ros2_java:dashing has all commits that osrf:dashing has

```
$ git rev-list cf2ef0707f7420f9b83e6c053775c9d52752a9e2^..origin/dashing
```

d4163f815c3e8f89b3eca9722eabf31245eb837d
4681ca3fc20971854938cdf816d8ad09178b67bc
a9244b8b72f24bb1eb7b7311d400ab1399f3718c
e5dcec25f0e7db86970328ab19df9c7981a1cc9f
0a3df77299a962d45192d39102df06b754f2ae2b
c3a2d7ad2ea7061a88914cadd744a5c85a6269b8
0e8e8cc8eda774836a3d93bccde6b7db1ea85b75
efe2df5df13d8ccc43d3408b1d36a3e0a0826f7c
66d622c8a55f61f847654e1b9d8a4152d6b21a13
0b48353d82a75e112d29f60305aee980e19c13de
720cbed528761c21ff1ee1e95d414f2e7cad7d30
7fb3dfe40689b84d3b74adc4ebfae97062804684
95fefa64b6348049c768af09678aa0bdcf7d4427
cf2ef0707f7420f9b83e6c053775c9d52752a9e2

\# i.e. There are 13 commits in ros2_java:dashing not in osrf:dashing

---

There are also 8 commits on osrf:dashing that aren't in osrf:galactic_devel, and there were some conflicts when rebasing :confused:.

jacobperron commented:
> I see some changes in https://github.com/ros2_java/ros2_java/commits/dashing that aren't here.

Oops, you're right. I forgot to include changes since this PR was opened. I've reapplied changes from `dashing`.

ivanpauno commented:
> Oops, you're right. I forgot to include changes since this PR was opened. I've reapplied changes from `dashing`.

I have checked locally, I can rebase `osrf:galactic_devel` on `ros2_java:dashing` more or less easily https://github.com/osrf/ros2_java/tree/ivanpauno/galactic-devel-rebase-origin-dashing.
So I think this fast-forward is a good idea :+1:.

---

This isn't actually a fast-forward though, and rebasing again on this branch adds more conflicts.
Can we just make the `dashing` branch the exact new `main`?

There isn't anything interesting in `main` that isn't in dashing really:

```
$ git log $(git merge-base origin/main origin/dashing)^..origin/main
commit b4e64f71538a057b3a1dbb49315aabeeedab3c88 (origin/master, origin/main, origin/HEAD)
Author: Esteve Fernandez <esteve@apache.org>
Date:   Fri Dec 18 07:45:04 2020 +0100

    Pin version of rcl_interfaces

commit f861280c504d311991c4152635778edea711592f
Author: Esteve Fernandez <esteve@apache.org>
Date:   Fri Dec 18 07:44:25 2020 +0100

    Pin version of rcl_interfaces

commit 62918477f4fb8810d4df0ac9c9e67b885b635d01 (osrf/master)
Author: Martin Huber <mhubii@users.noreply.github.com>
Date:   Fri Jan 17 14:46:25 2020 +0100

    Update README.md (#72)
```

The changes on the readme were already done in the `dashing` branch and the pinning is irrelevant.

jacobperron commented:
Good point. I'm fine just making this the new `main`, pending approval.

ivanpauno commented:
> Good point. I'm fine just making this the new `main`, pending approval.

To be more clear, the new `main` would be the [dashing branch](https://github.com/ros2-java/ros2_java/tree/dashing) and not this branch (the diff between the two branches is empty, it only makes the process of cherry-picking changes from `osrf:ros2_java` a bit easier to me.)

jacobperron commented:
> To be more clear, the new main would be the dashing branch and not this branch
Right, that's what I meant :+1:

jacobperron commented:
Alright, with all the comments addressed, I've created a new branch `main` based on the `dashing` branch and made that the default branch for the repository. We should start targeting new development there. I'll open up a PR documenting our procedure.
